File: app/src/main/java/com/example/pam_app/fragment/AddBucketEntryFragment.java
package com.example.pam_app.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.pam_app.R;
import com.example.pam_app.db.WallyDatabase;
import com.example.pam_app.model.BucketEntry;
import com.example.pam_app.presenter.AddBucketEntryPresenter;
import com.example.pam_app.repository.BucketMapper;
import com.example.pam_app.repository.BucketRepository;
import com.example.pam_app.repository.RoomBucketRepository;
import com.example.pam_app.utils.schedulers.AndroidSchedulerProvider;
import com.example.pam_app.utils.schedulers.SchedulerProvider;
import com.example.pam_app.view.AddBucketEntryView;
import com.google.android.material.datepicker.MaterialDatePicker;
import com.google.android.material.textfield.TextInputLayout;

import java.util.Calendar;
import java.util.List;
import java.util.TimeZone;

import static android.app.Activity.RESULT_OK;

public abstract class AddBucketEntryFragment extends Fragment implements AddBucketEntryView {

    public static final int MAX_AMOUNT = 1000000000;
    public static final int MAX_CHARACTERS = 50;
    private AddBucketEntryPresenter presenter;
    private View createdView;

    private EditText description;
    private EditText amount;
    private EditText selectedDate;
    private Calendar date;
    private AutoCompleteTextView bucket;
    private TextInputLayout dropdown;
    private MaterialDatePicker<Long> datePicker;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        final BucketRepository bucketRepository = new RoomBucketRepository(
                WallyDatabase.getInstance(requireActivity().getApplicationContext()).bucketDao(),
                new BucketMapper()
        );
        final SchedulerProvider schedulerProvider = new AndroidSchedulerProvider();
        presenter = new AddBucketEntryPresenter(this, bucketRepository, schedulerProvider);
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        createdView = view;
        presenter.onViewAttached();

        description = view.findViewById(R.id.description);
        amount = view.findViewById(R.id.amount);
        selectedDate = view.findViewById(R.id.date);
        date = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        bucket = view.findViewById(R.id.bucket);
        datePicker = MaterialDatePicker.Builder.datePicker().setTitleText(getString(R.string.pick_a_date)).build();
        dropdown = view.findViewById(R.id.bucket_dropdown);

        bucket.setText(getArguments().getString("bucket_name"), false);

        setDatePicker();
        setSaveEntryListener();
    }

    @Override
    public void setDropDownOptions(final List<String> buckets) {
        final ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(), R.layout.list_item, buckets);
        final AutoCompleteTextView editTextFilledExposedDropdown = createdView.findViewById(R.id.bucket);
        editTextFilledExposedDropdown.setAdapter(adapter);
    }

    @Override
    public void onErrorSavingBucketEntry() {
        Toast.makeText(
                getContext(),
                getString(R.string.entry_saving_failed),
                Toast.LENGTH_LONG
        ).show();
    }

    @Override
    public void onSuccessSavingBucketEntry(final BucketEntry entry) {
        Toast.makeText(
                getContext(),
                getString(R.string.entry_saving_success, entry.getComment()),
                Toast.LENGTH_LONG
        ).show();
        final Intent result = new Intent();
        result.putExtra("entry", entry);
        requireActivity().setResult(RESULT_OK, result);
        getActivity().finish();
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.onViewDetached();
    }

    @Override
    public void showDescriptionError(final int error, final Integer parameter) {
        if (parameter == null) {
            description.setError(getString(error));
        } else {
            description.setError(getString(error, parameter));
        }
    }

    @Override
    public void showAmountError(final int error, final Integer parameter) {
        if (parameter == null) {
            amount.setError(getString(error));
        } else {
            amount.setError(getString(error, parameter));
        }
    }

    @Override
    public void showDateError(final int error) {
        selectedDate.requestFocus();
        getParentFragmentManager().beginTransaction().remove(datePicker).commit();
        selectedDate.setError(getString(error));
    }

    @Override
    public void showBucketTitleError(int error) {
        dropdown.setError(getString(error));
    }


    private void setSaveEntryListener() {
        final Button saveEntry = createdView.findViewById(R.id.save);
        saveEntry.setOnClickListener(v -> presenter.saveBucketEntry(
                amount.getText().toString(),
                date.getTime(),
                description.getText().toString(),
                bucket.getText().toString()
        ));
    }

    private void setDatePicker() {
        datePicker.addOnPositiveButtonClickListener(selection -> {
            selectedDate.setText(datePicker.getHeaderText());
            date.setTimeInMillis(selection);
        });

        selectedDate.setOnFocusChangeListener((v, hasFocus) -> {
            if (hasFocus && !datePicker.isAdded()) {
                datePicker.show(getParentFragmentManager(), "date_picker");
            }
        });

        selectedDate.setOnClickListener(v -> {
            if (!datePicker.isAdded()) {
                datePicker.show(getParentFragmentManager(), "date_picker");
            }
        });
    }
}

File: app/src/main/java/com/example/pam_app/fragment/AddBucketEntryFragment.kt
package com.example.pam_app.fragment

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import com.example.pam_app.R
import com.example.pam_app.di.ContainerLocator
import com.example.pam_app.model.BucketEntry
import com.example.pam_app.presenter.AddBucketEntryPresenter
import com.example.pam_app.view.AddBucketEntryView
import com.google.android.material.datepicker.MaterialDatePicker
import com.google.android.material.textfield.TextInputLayout
import java.util.*

abstract class AddBucketEntryFragment : Fragment(), AddBucketEntryView {
    private var presenter: AddBucketEntryPresenter? = null
    private var createdView: View? = null
    private var description: EditText? = null
    private var amount: EditText? = null
    private var selectedDate: EditText? = null
    private var date: Calendar? = null
    private var bucket: AutoCompleteTextView? = null
    private var dropdown: TextInputLayout? = null
    private var datePicker: MaterialDatePicker<Long>? = null
    override fun onCreateView(inflater: LayoutInflater, viewGroup: ViewGroup?, savedInstanceState: Bundle?): View? {
        val container = ContainerLocator.locateComponent(context)
        presenter = AddBucketEntryPresenter(this, container?.bucketRepository,
                container?.schedulerProvider)
        return super.onCreateView(inflater, viewGroup, savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        createdView = view
        presenter!!.onViewAttached()
        description = view.findViewById(R.id.description)
        amount = view.findViewById(R.id.amount)
        selectedDate = view.findViewById(R.id.date)
        TimeZone.setDefault(TimeZone.getTimeZone("UTC"));
        date = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        bucket = view.findViewById(R.id.bucket)
        datePicker = MaterialDatePicker.Builder.datePicker().setTitleText(getString(R.string.pick_a_date)).build()
        dropdown = view.findViewById(R.id.bucket_dropdown)
        bucket?.setText(requireArguments().getString("bucket_name"), false)
        setDatePicker()
        setSaveEntryListener()
    }

    override fun setDropDownOptions(buckets: List<String?>?) {
        val adapter = ArrayAdapter(requireContext(), R.layout.list_item, buckets!!)
        val editTextFilledExposedDropdown = createdView!!.findViewById<AutoCompleteTextView>(R.id.bucket)
        editTextFilledExposedDropdown.setAdapter(adapter)
    }

    override fun onErrorSavingBucketEntry() {
        Toast.makeText(
                context,
                getString(R.string.entry_saving_failed),
                Toast.LENGTH_LONG
        ).show()
    }

    override fun onSuccessSavingBucketEntry(bucketEntry: BucketEntry) {
        Toast.makeText(
                context,
                getString(R.string.entry_saving_success, bucketEntry.comment),
                Toast.LENGTH_LONG
        ).show()
        val result = Intent()
        result.putExtra("entry", bucketEntry)
        requireActivity().setResult(Activity.RESULT_OK, result)
        requireActivity().finish()
    }

    override fun onStop() {
        super.onStop()
        presenter!!.onViewDetached()
    }

    override fun showDescriptionError(error: Int, parameter: Int?) {
        if (parameter == null) {
            description!!.error = getString(error)
        } else {
            description!!.error = getString(error, parameter)
        }
    }

    override fun showAmountError(error: Int, parameter: Int?) {
        if (parameter == null) {
            amount!!.error = getString(error)
        } else {
            amount!!.error = getString(error, parameter)
        }
    }

    override fun showDateError(error: Int) {
        selectedDate!!.requestFocus()
        parentFragmentManager.beginTransaction().remove(datePicker!!).commit()
        selectedDate!!.error = getString(error)
    }

    override fun showBucketTitleError(error: Int) {
        dropdown!!.error = getString(error)
    }

    private fun setSaveEntryListener() {
        val saveEntry = createdView!!.findViewById<Button>(R.id.save)
        saveEntry.setOnClickListener {
            presenter!!.saveBucketEntry(
                    amount!!.text.toString(),
                    if (selectedDate!!.text.toString() == "") null else date!!.time,
                    description!!.text.toString(),
                    bucket!!.text.toString()
            )
        }
    }

    private fun setDatePicker() {
        datePicker!!.addOnPositiveButtonClickListener { selection: Long? ->
            selectedDate!!.setText(datePicker!!.headerText)
            date!!.timeInMillis = selection!!
        }
        selectedDate!!.onFocusChangeListener = OnFocusChangeListener { _: View?, hasFocus: Boolean ->
            if (hasFocus && !datePicker!!.isAdded) {
                datePicker!!.show(parentFragmentManager, "date_picker")
            }
        }
        selectedDate!!.setOnClickListener {
            if (!datePicker!!.isAdded) {
                datePicker!!.show(parentFragmentManager, "date_picker")
            }
        }
    }

    companion object {
        const val MAX_AMOUNT = 1000000000
        const val MAX_CHARACTERS = 50
    }
}